### PR TITLE
[SPARK-12804][ML] Fix LogisticRegression with FitIntercept on all same label training data

### PR DESCRIPTION
**Pull request by feynmanliang**
CC @jkbradley @mengxr @dbtsai

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 1]
```diff
@@ -339,9 +339,11 @@ class LogisticRegression @Since("1.2.0") (
          b = \log{P(1) / P(0)} = \log{count_1 / count_0}
          }}}
        */
-      initialCoefficientsWithIntercept.toArray(numFeatures)
-        = math.log(histogram(1) / histogram(0))
-    }
+       if (histogram.length >= 2) { // check to make sure indexing into histogram(1) is safe
+         initialCoefficientsWithIntercept.toArray(numFeatures) = math.log(
+           histogram(1) / histogram(0))
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: In this case, the whole training step can be skipped. Currently, we only support binary LoR, so the max of `histogram.length` will be two. In LiR, when the `yStd == 0.0`, the model will be returned immediately without training, see https://github.com/feynmanliang/spark/blob/SPARK-12804/mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala#L226

We can do similar thing here like

```scala
if (histogram.length == 2) {
  if (histogram(0) == 0.0) {
    model = (new LogisticRegressionModel(uid, Vectors.sparse(numFeatures, Seq()), Double.PositiveInfinity))
    return model
  } else {
    initialCoefficientsWithIntercept.toArray(numFeatures) = math.log(
            histogram(1) / histogram(0))
  } else if (histogram.length == 1) {
    model = (new LogisticRegressionModel(uid, Vectors.sparse(numFeatures, Seq()), Double.NegativeInfinity))
    return model
  } else {
    some excpetion
  }
}
```

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 2]
```diff
@@ -883,6 +884,22 @@ class LogisticRegressionSuite
...
+    val sameLabels = dataset
+      .withColumn("zeroLabel", lit(0.0))
+      .withColumn("oneLabel", lit(1.0))
+
+    val model = lr
+      .setLabelCol("oneLabel")
+      .fit(sameLabels)
+
+    assert(model.coefficients ~== Vectors.dense(0.0) absTol 1E-3)
+    assert(model.intercept === Double.PositiveInfinity)
```
Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: Can you add another test showing that all `zeroLabel` will return intercept with `Double.NegativeInfinity`?

Thanks.

Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: BTW, this bug should not happen when all the labels are one since the histogram should be still size of two.

Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: Thanks for pointing that out!

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 1 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Thank you!

**Comment by feynmanliang** [comment]
@dbtsai @jkbradley ready for second review.

The big diff is because I grouped the same label cases with the normal case to generate `coefficients`, `intercept`, and `objectiveTrace` all in the same block. This is to avoid repeated code when generating the model summary.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 3]
```diff
@@ -883,6 +884,27 @@ class LogisticRegressionSuite
...
+    assert(allZeroModel.coefficients ~== Vectors.dense(0.0) absTol 1E-3)
+    assert(allZeroModel.intercept === Double.NegativeInfinity)
+
+    val allOneModel = lr
+      .setLabelCol("oneLabel")
+      .fit(sameLabels)
+    assert(allOneModel.coefficients ~== Vectors.dense(0.0) absTol 1E-3)
+    assert(allOneModel.intercept === Double.PositiveInfinity)
+  }
+
```
Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: Can you add one test which is all labels the same but `fitIntercept=false` here to avoid the issue in LiR? Thanks.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 4]
```diff
@@ -883,6 +884,27 @@ class LogisticRegressionSuite
     assert(model1a0.intercept ~== model1b.intercept absTol 1E-3)
   }
 
+  test("logistic regression with fitIntercept=true and all labels the same") {
+    val lr = new LogisticRegression()
+      .setFitIntercept(true)
+      .setMaxIter(3)
```
Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: add extra line.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 5]
```diff
@@ -276,113 +276,123 @@ class LogisticRegression @Since("1.2.0") (
...
+          s"binary classification. Found $numClasses in the input dataset."
+        logError(msg)
+        throw new SparkException(msg)
+      } else if ($(fitIntercept) && numClasses == 2 && histogram(0) == 0.0) {
+        logWarning(s"All labels are one and fitIntercept=true, so the coefficients will be " +
+          s"zeros and the intercept will be positive infinity; as a result, " +
+          s"training is not needed.")
+        (Vectors.sparse(numFeatures, Seq()), Double.PositiveInfinity, Array.empty[Double])
+      } else if ($(fitIntercept) && numClasses == 1) {
+        logWarning(s"All labels are one and fitIntercept=true, so the coefficients will be " +
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: "All labels are zero...."

**Comment by dbtsai** [comment]
LGTM except couple minor issues. Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 6]
```diff
@@ -276,113 +276,123 @@ class LogisticRegression @Since("1.2.0") (
...
-        = math.log(histogram(1) / histogram(0))
-    }
 
-    val states = optimizer.iterations(new CachedDiffFunction(costFun),
-      initialCoefficientsWithIntercept.toBreeze.toDenseVector)
+        val initialCoefficientsWithIntercept =
+          Vectors.zeros(if ($(fitIntercept)) numFeatures + 1 else numFeatures)
+
+        if ($(fitIntercept)) {
+          /*
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Indentation in the comment is off.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala (hunk 3 above)
Comment (dbtsai) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: Also, check if objectiveHistory has length of zero.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 5 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: OK

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 6 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: OK

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala (hunk 3 above)
Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: OK

**Comment by feynmanliang** [comment]
@dbtsai Added `fitIntercept=false` tests and fixed comments/`logWarning` messages.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 6 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Still off. Add one char in the following line.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 7]
```diff
@@ -276,113 +276,123 @@ class LogisticRegression @Since("1.2.0") (
...
+            it follows the distribution of the labels.
+
+            {{{
+              P(0) = 1 / (1 + \exp(b)), and
+              P(1) = \exp(b) / (1 + \exp(b))
+            }}}, hence
+            {{{
+              b = \log{P(1) / P(0)} = \log{count_1 / count_0}
+            }}}
+          */
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: add one space before `*/`

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 6 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Sorry >.<, fixed

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 7 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: OK

**Comment by feynmanliang** [comment]
@dbtsai validating coefficients with R will be harder than expected, `glmnet` requires feature dimension >= 2 and `glm` doesn't yield +/- Infinity intercepts when given all 0/1 datasets...

**Comment by dbtsai** [comment]
Ideally, we would like to see the one without intercept to ensure that nothing will not be messed up during the refactoring. I'll merge now, and let's add the test for all same labels without intercept. Thanks.

**Comment by dbtsai** [comment]
Merged into master. Thanks.

**Comment by dbtsai** [comment]
A followup PR is created: https://issues.apache.org/jira/browse/SPARK-12908
